internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -238,38 +242,20 @@ func (r *AgentServiceConfigReconciler) Reconcile(origCtx context.Context, req ct
 		{"WebHookClusterRoleBinding", aiv1beta1.ReasonWebHookClusterRoleBindingFailure, r.newWebHookClusterRoleBinding},
 		{"WebHookAPIService", aiv1beta1.ReasonWebHookAPIServiceFailure, r.newWebHookAPIService},
 	} {
-		obj, mutateFn, err := component.fn(ctx, log, instance)
-		if err != nil {
-			msg := "Failed to generate definition for " + component.name
-			log.WithError(err).Error(msg)
-			conditionsv1.SetStatusConditionNoHeartbeat(&instance.Status.Conditions, conditionsv1.Condition{
-				Type:    aiv1beta1.ConditionReconcileCompleted,
-				Status:  corev1.ConditionFalse,
-				Reason:  component.reason,
-				Message: msg,
-			})
-			if statusErr := r.Status().Update(ctx, instance); statusErr != nil {
-				log.WithError(err).Error("Failed to update status")
-				return ctrl.Result{Requeue: true}, statusErr
-			}
-			return ctrl.Result{Requeue: true}, err
+		if result, err := r.reconcileComponent(ctx, instance, log, component); err != nil {
+			return result, err
 		}
+	}
 
-		if result, err := controllerutil.CreateOrUpdate(ctx, r.Client, obj, mutateFn); err != nil {
-			msg := "Failed to ensure " + component.name
-			log.WithError(err).Error(msg)
-			conditionsv1.SetStatusConditionNoHeartbeat(&instance.Status.Conditions, conditionsv1.Condition{
-				Type:    aiv1beta1.ConditionReconcileCompleted,
-				Status:  corev1.ConditionFalse,
-				Reason:  component.reason,
-				Message: msg,
-			})
-			if statusErr := r.Status().Update(ctx, instance); statusErr != nil {
-				log.WithError(err).Error("Failed to update status")
-				return ctrl.Result{Requeue: true}, statusErr
+	// Additional routes need to be synced if HTTP iPXE routes are exposed
```

Comment:
Is there reconcile logic to ensure that these do not exist in case it's set to disabled? I may have missed that, but didn't see it on my first read-through.

Reply:
Yes, we have unit tests:
* "should not create plain http route by default"
* "should not create plain http route if ExposeIPXEHTTPRoute is explicitly disabled"
* "should not create plain http route if ExposeIPXEHTTPRoute is not unknown"

Reply:
I am looking for:

- should remove plain http route if ExposeIPXEHTTPRoute changes from "enabled" to "disabled"

Do we have that?

Reply:
Ah, good idea, added that - `should remove http route after IPXEHTTPRouteEnabled changed to disabled`.

However that won't update links in InfraEnv - this would need to be fixed in a separate PR I suppose

Reply:
Why defer fixing the infraenv links?

Speaking of which: Don't we want a separate set of non-TLS links on each infraenv, so we can keep the TLS links as well? If a user wants to do non-TLS ipxe in one infranev, that doesn't necessarily mean they want to turn of TLS for all boot artifacts everywhere, right? Or are we assuming that `InfraEnv.status.bootArtifacts` is only ever used for ipxe, and thus we think it's reasonable to treat TLS as an all-or-nothing setting?

Reply:
>Why defer fixing the infraenv links?

There are two scenarios when iPXE is being used:
* http links are used directly - dhcp server is configured to consume ipxeScript straight from ACM
* https links are used indirectly - customer fetches the artifacts, places them on a web server manually etc.
This might be the case for installations where ACM is not reachable by DHCP server directly, ipxe scripts need modifications (i.e. kernel args) etc.

As a result this setting is used to simplify ipxe script access in customer environment, not as a preference or customization. Its not something people want to change throughout the hub's lifecycle - if a more complicated system is required, ipxeScript would need to be adjusted manually anyway, so http/https access is orthogonal

>If a user wants to do non-TLS ipxe in one infranev, that doesn't necessarily mean they want to turn of TLS for all boot artifacts everywhere, right? Or are we assuming that InfraEnv.status.bootArtifacts is only ever used for ipxe, and thus we think it's reasonable to treat TLS as an all-or-nothing setting?

`boot-artifacts` are standard RHCOS kernel/rootfs, so these are

Reply:
Maybe it would help to clarify what exactly you meant by "that won't update links in InfraEnv - this would need to be fixed in a separate PR". What update won't happen? Is it that the links in `bootArtifacts` would all be http, and remain that way without https?

> As a result this setting is used to simplify ipxe script access in customer environment, not as a preference or customization.

"Setting", "preference", "customization", I'm not sure what the significance is here of how they are different?

> Its not something people want to change throughout the hub's lifecycle

Given hardware refresh cycles, I fully expect management clusters to out-live the hardware they are managing. Or to be managing multiple generations of hardware at the same time. Some that use ipxe, some that use redfish virtualmedia perhaps, etc.

Not to mention, this is a declarative API. Whatever desired state is specified should be achievable regardless of current real-world state.

> boot-artifacts are standard RHCOS kernel/rootfs, so these are

Are there words missing at the end?

Reply:
> What update won't happen? Is it that the links in bootArtifacts would all be http, and remain that way without https?

As far as I understand this will generally work as we expect it to. The only issue being that we don't update any of the boot artifact urls unless the iso url changes. This was done to avoid constant churn on the auth token portion of the URL and I feel like changing this behavior is out of scope for this PR.

If the user changes the `ExposeIPXEHTTPRoute` we will rollout the assisted service deployment and whenever the existing logic says to update an infraEnv it will be updated with the correct URL scheme.

> Don't we want a separate set of non-TLS links on each infraenv, so we can keep the TLS links as well? 

I'm not sure this is important. The individual boot artifact URLs were only exposed to allow ironic to fetch the artifacts individually rather than requiring them to pull the artifacts out of our iPXE boot script (they provide their own iPXE script to boot hosts in the converged ZTP flow). If a user has mixed deployment methods then the ISO link will always be https, and the ipxe script url scheme will depend on the env var and whether it was updated as a described above.

That all said I do think we need to improve the infraEnv boot artifact handling eventually. I could see a bug where someone had http exposed, removed it, then ZTP starts failing because the scheme wasn't updated in the infraEnv, but I think that's an uncommon enough case (not to mention that ZTP isn't even using ipxe yet) that we can handle it in a followup.

Reply:
updated the logic (initrd URL is updated and so are other links) and added a unittest